src/main/java/duke/parser/Parser.java

```diff
@@ -0,0 +1,195 @@
+
+import duke.commands.BaseCommand;
+import duke.commands.ExitCommand;
+import duke.commands.tasks.AddTaskCommand;
+import duke.commands.tasks.DeleteTaskCommand;
+import duke.commands.tasks.FindTaskCommand;
+import duke.commands.tasks.ListTasksAfterCommand;
+import duke.commands.tasks.ListTasksBeforeCommand;
+import duke.commands.tasks.ListTasksCommand;
+import duke.commands.tasks.MarkTaskCommand;
+import duke.commands.tasks.UnmarkTaskCommand;
+import duke.domain.Task;
+import duke.domain.Todo;
+import duke.exceptions.IncorrectArgumentException;
+import duke.exceptions.InvalidDateTimeException;
+import duke.exceptions.InvalidTaskSpecificationException;
+import duke.exceptions.MissingArgumentException;
+import duke.exceptions.NoCommandException;
+import duke.exceptions.UnknownCommandException;
+
+/**
+ * Parser Class in charge of parsing the users' command.
+ */
+public class Parser {
+
+    private static final String listCommand = "list";
+    private static final String markCommand = "mark";
+    private static final String todoCommand = "todo";
+    private static final String deadlineCommand = "deadline";
+    private static final String findCommand = "find";
+    private static final String deadlineSubCommand = " /by ";
+    private static final String eventCommand = "event";
+    private static final String eventSubCommand = " /at ";
+    private static final String advancedListSubCommand1 = "/before ";
+    private static final String advancedListSubCommand2 = "/after ";
+    private static final String unmarkCommand = "unmark";
+    private static final String deleteCommand = "delete";
+    private static final String exitCommand = "bye";
+    private static final DateTimeFormatter formatter = DateTimeFormatter.ofPattern(
+            "dd-MM-yyyy HH:mm");
+
+    private static LocalDateTime parseInputString(String inputDateTime)
+            throws DateTimeParseException {
+        return LocalDateTime.parse(inputDateTime, formatter);
+    }
+
+    /**
+     * Returns a boolean corresponding to whether the given string is numeric.
+     *
+     * @param strNum
+     *            String to be tested
+     * @return boolean representing whether the string is numeric
+     */
+    private static boolean isNumeric(String strNum) {
+        if (strNum == null) {
+            return false;
+        }
+        try {
+            Integer.parseInt(strNum);
+        } catch (NumberFormatException nfe) {
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * The parse function takes a string as input and parses it into commands.
+     * The parse function is responsible for parsing the user's input into
+     * commands that can be executed by the program. It also handles any errors
+     * that may occur during parsing, such as missing arguments or invalid date/time
+     * formats.
+     *
+     * @param userInput
+     *            Get the user input
+     */
+    public BaseCommand parse(String userInput)
```

Comment:
Maybe can find a way to shorten this method? eg : put some of the operation in separate method and call it in here?

Reply:
This is definitely one area of my code that could use refactoring, I will take note of this in the future